euphonic/fast_adaptive_broadening.py

```diff
@@ -0,0 +1,166 @@
+"""
+Functions for fast adaptive broadening of density of states spectra
+"""
+from typing import Optional
+from scipy.optimize import curve_fit
+from scipy.signal import convolve
+import numpy as np
+from euphonic import Quantity
+
+
+def fast_broaden(dos_bins: np.ndarray,
+                 freqs: np.ndarray,
+                 mode_widths: Quantity,
+                 weights: np.ndarray,
+                 mode_weights: np.ndarray,
+                 adaptive_error: float) -> np.ndarray:
+    """
+    Uses a fast, approximate method to adaptively broaden a density
+    of states spectrum
+
+    Parameters
+    ----------
+    dos_bins
+        Shape (n_e_bins + 1,) float ndarray. The energy bin edges to use for
+        calculating the DOS
+    freqs
+        Shape (n_qpts, n_modes) float ndarray. Frequencies per q-point
+        and mode.
+    mode_widths
+        Shape (n_qpts, n_modes) float Quantity in energy units. The broadening
+        width for each mode at each q-point
+    weights
+        Shape (n_qpts,) float ndarray. The weight for each q-point.
+    mode_weights
+        Shape (n_qpts, n_modes) float ndarray. The weight of each mode at
+        each q-point.
+    adaptive_error
+        Scalar float. Acceptable error for gaussian approximations, defined
+        as the absolute difference between the areas of the true and
+        approximate gaussians.
+
+    Returns
+    -------
+    dos
+        Float ndarray of shape (dos_bins - 1,) containing density of states
+        ydata
+    """
+    n_modes = freqs.shape[-1]
+    freqs = np.ravel(freqs)
+    mode_weights = np.ravel(mode_weights)
+    mode_widths = np.ravel(mode_widths).to('hartree').magnitude
+    n_weights = np.repeat(weights, n_modes)
+
+    # set default error value if none specified
+    if adaptive_error is None:
+        adaptive_error = 0.01
+
+    # determine spacing value for mode_width samples given desired error level
+    spacing = 656.1*adaptive_error**3 - 131.8*adaptive_error**2 +\
+                15.98*adaptive_error + 1.0803
+
+    bin_width = dos_bins[1]-dos_bins[0]
+    mode_widths = np.maximum(mode_widths, bin_width/2)
+
+    n_kernels = int(
+        np.ceil(np.log(max(mode_widths)/min(mode_widths))/np.log(spacing)))
+    mode_width_samples = spacing**np.arange(n_kernels+1)*min(mode_widths)
+    freq_range = 3*max(mode_widths)
+    kernel_npts_oneside = np.ceil(freq_range/bin_width)
+    kernels = gaussian(xvals=np.arange(-kernel_npts_oneside,
+                       kernel_npts_oneside+1, 1)*bin_width,
+                       sigma=mode_width_samples[:, np.newaxis],
+                       centre=0)*bin_width
+    kernels_idx = np.searchsorted(mode_width_samples, mode_widths)
+
+    lower_coeffs = find_coeffs(spacing)
+    scaled_data_matrix = np.zeros((len(dos_bins)-1, len(kernels)))
+    for i in range(1, len(mode_width_samples)):
```

Comment:
Might be worth a comment to explain that we intentionally skip points with insert-position 0 because they lie outside the bin range.